bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/OnStackMarking.scala

```diff
@@ -0,0 +1,221 @@
+   */
+  def affectedVariables(affectedSet: mutable.Set[Variable], marking: (OnStackMarking)=>Unit): mutable.Set[Variable] = {
+    val affected = if (affectedSet == null) mutable.Set[Variable]() else affectedSet
+    val marker = new OnStackMarking(declVars.clone(), Some(affected)) {
+      // do nothing when encountering a "bond on stack" node. Bound and Used variables will get the same treatment
+      override def onStackBind(bind: BindCommon, v: Variable) = {}
+    }
+    marking(marker)
+    affected
+  }
+  
+  /**
+   * Runs the algorithm in sandboxed branches and merges the result in current marker
+   */
+  def conjugate[A](collector: mutable.HashMap[Symbol, VarAndState], branches: ((OnStackMarking)=>Unit)*): Unit = {
+    val collectorClone = collector.clone()
+    val modified = mutable.Set[Variable]()
+    val bindNodes = mutable.HashMap[Symbol, Seq[BindCommon]]()
+    for ((branch, i) <- branches.zipWithIndex) {
+      val current = if (i == 0) collector else collectorClone.clone() 
+    	val marker = new OnStackMarking(current, Some(modified)) {
+        // In the sandboxed environment, we collect "bind on stack" nodes to defer decision
+    	  override def onStackBind(bind: BindCommon, v: Variable) =
+    	    bindNodes.put(v.symbol, bindNodes.getOrElse(v.symbol, Seq()) :+ bind)
+    	}
+      branch(marker)
+      if (i != 0) {
+        for (v <- modified) {
+          val collectorVS = collector.get(v.symbol).get
+          val currentVS = current.get(v.symbol).get
+          if (collectorVS.state != currentVS.state && collectorVS.state != ToInstantiate) {
+            collector.put(v.symbol, VarAndState(collectorVS.v, ToInstantiate))
+          }
+        }
+      }
+    }
+    this.modified.foreach { _ ++= modified }
+    for ((sym, bindNodes) <- bindNodes) {
+      val vas = collector.get(sym).get
+      if (vas.state == BoundFirst) {
+        for (bind <- bindNodes) {
+          // Now, we confirm the fact those bindNodes can work on the stack
+          onStackBind(bind, vas.v)
+        }
+      }
+    }
+  }
+  
+  /**
+   * The toplevel OnStack markers modify the bind statements and expressions
+   */
+  def onStackBind(bind: BindCommon, v: Variable) = {
+    bind.onStack = true
+  }
+  
+  def bindBindCommon(bind: BindCommon, sym: Symbol, right: Expression) = {
+		  walkExpr(right)
+		  // If we bind and it has not been used, put it on stack
+		  for (decl <- declVars.get(sym) if !decl.state.locked) {
+		    tagVariable(decl.v, BoundFirst)
+		    onStackBind(bind, decl.v)
+		  }
+  }
+  
+  override def walkStat(statement: Statement): Unit = statement match {
+    case bind @ BindStatement(left @ Variable(sym), right) =>
+      bindBindCommon(bind, sym, right)
+      
+    case CallStatement(callable, args) =>
+      walkExpr(callable)
+      args map walkExpr
+      for ((sym, decl) <- declVars if decl.state == Captured) {
+        tagVariable(decl.v, ToInstantiate)
+      }
```

Comment:
I see, so this instantiates conservatively all the captured variables.
Sounds good for now as otherwise you would need to follow which `proc` goes into which variable(s) and might end up as one of the arguments of this call.